CPAC/utils/utils.py

```diff
@@ -2657,3 +2657,192 @@ def flip_orientation_code(code):
+        case _:
+            return None
+
+    return None
+
+
+def get_fmap_metadata_at_build_time(sub_dict, orig_key, input_creds_path, dl_dir):
+    """Extract fieldmap metadata during workflow build time.
+
+    Parameters
+    ----------
+    sub_dict : dict
+        Subject dictionary containing fieldmap information
+    orig_key : str
+        Original fieldmap key name
+    input_creds_path : str
+        Path to AWS credentials
+    dl_dir : str
+        Download directory path
+
+    Returns
+    -------
+    dict
+        Dictionary containing fieldmap metadata, or None if unavailable
+    """
+    import json
+    import os
+
+    try:
+        # Check if scan_parameters exists for this fieldmap
+        if orig_key not in sub_dict["fmap"]:
+            return None
+
+        if "scan_parameters" not in sub_dict["fmap"][orig_key]:
+            return None
+
+        scan_params_path = sub_dict["fmap"][orig_key]["scan_parameters"]
+
+        # Handle dictionary metadata (direct dict)
+        if isinstance(scan_params_path, dict):
+            return scan_params_path
+
+        # Handle file path metadata
+        if isinstance(scan_params_path, str):
+            local_path = scan_params_path
+
+            # Handle S3 paths
+            if scan_params_path.startswith("s3://"):
+                try:
+                    local_path = check_for_s3(
+                        scan_params_path, input_creds_path, dl_dir
+                    )
+                except Exception:
+                    return None
+
+            # Load JSON file
+            if local_path.endswith(".json") and os.path.exists(local_path):
+                with open(local_path, "r", encoding="utf-8") as f:
+                    return json.load(f)
+
+    except (FileNotFoundError, json.JSONDecodeError, KeyError, Exception):
+        pass
+
+    return None
+
+
+def get_fmap_build_info(metadata_dict):
+    """Determine fieldmap processing requirements at build time.
+
+    Parameters
+    ----------
+    metadata_dict : dict or None
+        Fieldmap metadata dictionary
+
+    Returns
+    -------
+    dict
+        Dictionary with keys:
+        - 'fmap_type': str or None
+        - 'needs_echo_times': bool
+        - 'needs_phasediff_processing': bool
+        - 'is_epi': bool
+
+    Raises
+    ------
+    ValueError
+        If metadata_dict is None or if fieldmap type cannot be determined
+    """
+    from CPAC.utils.utils import get_fmap_type
+
+    if not metadata_dict:
+        raise ValueError(
+            "Fieldmap metadata dictionary is required but was None. "
+            "Cannot determine fieldmap processing requirements without metadata."
+        )
+
+    fmap_type = get_fmap_type(metadata_dict)
+
+    if fmap_type is None:
+        raise ValueError(
+            f"Could not determine fieldmap type from metadata: {metadata_dict}. "
+            "Metadata must contain required BIDS fields for fieldmap type detection."
+        )
```

Comment:
For the exceptions with f-string messages, like this one, [we should assign the message to a variable first, then pass that variable to the exception constructor](https://docs.astral.sh/ruff/rules/f-string-in-exception).

```suggestion
        msg = (f"Could not determine fieldmap type from metadata: {metadata_dict}. "
               "Metadata must contain required BIDS fields for fieldmap type detection.")
        raise ValueError(msg)
```